finish initial support for probabilistic forecast reports

wholmgren opened the pull request:
- [x] Closes #266  .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Reports work with `ProbabilisticForecast` and `ProbabilisticForecastConstantValue` with `axis='y'`, though they are not pretty. Smaller legend text would help.

I think, but haven't tested, that reports will work with `axis='x'` but the plots will be mislabeled.

I've not fixed the links in the pdf report.

wholmgren commented:
This [gist](https://gist.github.com/wholmgren/dee5ba8bee170a2eee1177d3f3ade0c8) shows a time series plot with a mix of `ProbabilisticForecast` and `ProbabilisticForecastConstantValue` on both the x and y axis. 

The y axis lines are good for `ProbabilisticForecastConstantValue` but a `ProbabilisticForecast` should probably have filled areas. That would require propagating even more metadata in `timeseries_meta_df`. An alternative might be building different `timeseries_value_df, timeseries_meta_df` for each set of obs, `ProbabilisticForecastConstantValue`, and `ProbabilisticForecats` in a report, and then passing those to `plotly_figures.timeseries` with a few more kwargs. In any case, it also requires fighting plotly's fill feature so I'm -1 on it for initial implementation.

Maybe (likely) there is a better way plot the x forecasts on the time series plot, but I can't think of it. Maybe we should stop trying to cram it into a single plot and we need a separate plot for probability vs time. This would require modifications to `plotly_figures.timeseries_plots` and the templates to make the 3rd plot optional.

wholmgren commented:
@alorenzo175 feedback on the possible directions of this PR would be useful at this point.

alorenzo175 commented:
That plot is very hard for me to understand. Definitely would be helpful to separate the 'x' axis ones into a probability vs time

wholmgren commented:
PR now separates plots into obs/deterministic/y-axis-prob

![newplot(21)](https://user-images.githubusercontent.com/4383303/85496182-c566fb00-b590-11ea-9a73-04dd21bfdf60.png)

and x-axis-prob

![newplot(20)](https://user-images.githubusercontent.com/4383303/85496212-cc8e0900-b590-11ea-8f58-610ed1a6ab79.png)

wholmgren commented:
@alorenzo175 this is ready for a review. I'm avoiding modifications to the pdf reports until the html is figured out.

wholmgren commented:
The [gist](https://gist.github.com/wholmgren/dee5ba8bee170a2eee1177d3f3ade0c8#file-test_xy_prob-html) now has an html file with the rendered report for a set of forecasts that includes axis = x and y.

<img width="586" alt="Screen Shot 2020-06-24 at 2 14 09 PM" src="https://user-images.githubusercontent.com/4383303/85628431-fe50af80-b624-11ea-8505-5e8df5037602.png">

alorenzo175 commented on solarforecastarbiter/reports/template.py:
```diff
@@ -89,6 +89,8 @@ def _get_render_kwargs(report, dash_url, with_timeseries):
             timeseries_specs = ('{}', '{}')
```
need to add a 3rd element for [2] below

wholmgren commented:
PDF links now work (tested locally).

I added a couple of tests to `test_cli.py` for probabilistic forecasts. The first has only forecasts on y axis - this makes a report with just the variable vs time plot and the scatter plot. The second has forecasts on both x and y - this adds the probability vs time plot. I confirmed that they render correctly by inspecting the output of the test functions. I could add a test to `test_plotly_figures` for the json/None behavior of `timeseries` but that function is not currently explicitly tested, probably because it relies on a report object that is a pain to construct.

The untested line in the diff was not tested before this PR.

wholmgren commented:
Added a new test for `timeseries_plots` to check that `str, str, None` is returned for the `conftest.raw_report`. Do we need a test for `str, str, str` based on a new `conftest.raw_report_xy` or some other method? Also added a new test for the naming function.

alorenzo175 commented on solarforecastarbiter/reports/templates/pdf/macros.j2:
```diff
@@ -115,7 +115,7 @@
     \endlastfoot
     %- for pfxobs in report.raw_report.processed_forecasts_observations
-    \VAR{pfxobs.name} &
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")} &
```
why not use the html_to_tex filter?

wholmgren replied:
ah, I did not understand until now that it's made accessible for calling within a jinja variable expression.

alorenzo175 commented on solarforecastarbiter/reports/templates/pdf/macros.j2:
```diff
@@ -125,11 +125,23 @@
     %- endif
     \VAR{pfxobs.original.data_object.interval_label} &
     \VAR{pfxobs.original.data_object.interval_length.total_seconds()//60 | int} min &
+    %- if pfxobs.original.forecast.constant_values is defined
```
macro?

alorenzo175 replied:
Oh I guess i don't know about macro within macro

wholmgren replied:
I looked at it for a bit but had a hard time with it

alorenzo175 commented on solarforecastarbiter/reports/templates/pdf/macros.j2:
```diff
@@ -187,7 +199,7 @@
     \endlastfoot
     %- for pfxobs in proc_fxobs_list
-    \VAR{pfxobs.name} & \VAR{pfxobs.original.data_object.name}
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")} & \VAR{pfxobs.original.data_object.name}
```
filter?

alorenzo175 commented on solarforecastarbiter/reports/templates/pdf/macros.j2:
```diff
@@ -245,7 +257,7 @@
     \endlastfoot
     %- for pfxobs in proc_fxobs_list
-    \VAR{pfxobs.name}
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")}
```
filter?

alorenzo175 commented on solarforecastarbiter/tests/test_cli.py:
```diff
@@ -328,6 +328,45 @@ def test_report_pdf(cli_token, mocker, report_objects, remove_orca):
+        infile = tmpdir + '/report.json'
+        with open(infile, 'w') as f:
+            f.write('{}')
+        outfile = tmpdir + '/test_out.html'
```
do both pdf and html?

alorenzo175 commented:
> Added a new test for `timeseries_plots` to check that `str, str, None` is returned for the `conftest.raw_report`. Do we need a test for `str, str, str` based on a new `conftest.raw_report_xy` or some other method? 

I guess it wouldn't hurt?

wholmgren commented:
@alorenzo175 here's the last test. this copy pasta pairs nicely with an IPA.

alorenzo175 commented:
good with me
